docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,456 @@
+
+The operation:
+
+- Is division modulo `2^32` to provide the remainder of the division when a number is divided by `2^32`.
+- Decomposes a Field element in the range `[0, 2^64]` into two 32-bit limbs, `remainder` and `quotient`, using this equation: `n = quotient * 2^32 + remainder`.
+- Acts as a proof that the input is in the range [0, 2^64).
+- Asserts that both `remainder` and `quotient` are in the range `[0, 2^32]` using [rangeCheck32()](#rangeCheck32).
+
+Example:
+
+```ts
+let n = Field((1n << 32n) + 8n)
+let { remainder, quotient } = Gadgets.divMod32(n);
+// remainder = 8, quotient = 1
+
+n.assertEquals(quotient.mul(1n << 32n).add(remainder));
+```
+
+## leftShift32()
+
+```ts
+leftShift32(field: Field, bits: number) => Field
+```
+
+The `leftShift32()` gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the left. Unlike rotation, the bits that fall off at the end are discarded and the vacant positions are filled with zeros.
+
+The operation:
+
+- Performs a left shift operation on a Field element.
+- Shifts bits to the left and discards the overflowing bits.
+- Performs these operations with the big-endian 32-bit representation of the number, where the most significant (32nd) bit is on the left end and the least significant bit is on the right end.
+- Requires values that are range checked to 32 bits. You can use [rangeCheck32](#rangecheck32).
+
+Example:
+
+```ts
+const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
+const y = Gadgets.leftShift32(x, 2); // left shift by 2 bits
+y.assertEquals(0b110000); // 48 in binary
+```
+
+## leftShift64()
+
+```ts
+leftShift64(field: Field, bits: number) => Field
+```
+
+The `leftShift64()` gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the left. Unlike rotation, the bits that fall off at the end are discarded and the vacant positions are filled with zeros.
+
+The operation:
+
+- Performs a left shift operation on a Field element.
+- Shifts bits to the left and discards the overflowing bits.
+- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
+- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).
+
+Example:
+
+```ts
+const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
+const y = Gadgets.leftShift(x, 2); // left shift by 2 bits
+y.assertEquals(0b110000); // 48 in binary
+
+const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
+leftShift(xLarge, 32); // throws an error since input exceeds 64 bits
+``
+
+## leftShift32()
```

Comment:
```suggestion
## rightShift32()
```

Looks like a whole lot of copy-pasta. I'll provide inline suggestions to correct on the assumption that it's the same but right-leaning.